src/worker/tasks/nonceResyncWorker.ts

```diff
@@ -0,0 +1,103 @@
+import { Job, Processor, Worker } from "bullmq";
+import { eth_getTransactionCount, getRpcClient } from "thirdweb";
+import {
+  lastUsedNonceKey,
+  recycleNonce,
+  splitSentNoncesKey,
+} from "../../db/wallets/walletNonce";
+import { getConfig } from "../../utils/cache/getConfig";
+import { getChain } from "../../utils/chain";
+import { logger } from "../../utils/logger";
+import { redis } from "../../utils/redis/redis";
+import { thirdwebClient } from "../../utils/sdk";
+import { NonceResyncQueue } from "../queues/nonceResyncQueue";
+import { logWorkerExceptions } from "../queues/queues";
+
+// Must be explicitly called for the worker to run on this host.
+export const initNonceResyncWorker = async () => {
+  const config = await getConfig();
+  if (config.minedTxListenerCronSchedule) {
+    NonceResyncQueue.q.add("cron", "", {
+      repeat: { pattern: config.minedTxListenerCronSchedule },
+      jobId: "nonce-resync-cron",
+    });
+  }
+
+  const _worker = new Worker(NonceResyncQueue.q.name, handler, {
+    connection: redis,
+    concurrency: 1,
+  });
+  logWorkerExceptions(_worker);
+};
+
+export const sleep = async (ms: number) =>
+  new Promise((resolve) => setTimeout(resolve, ms, null));
+
+/**
+ * Resyncs nonces for all wallets.
+ * This worker should be run periodically to ensure that nonces are not skipped.
+ * It checks the onchain nonce for each wallet and recycles any missing nonces.
+ */
+const handler: Processor<any, void, string> = async (job: Job<string>) => {
+  const sentNoncesKeys = await redis.keys("*nonce-sent*");
+  job.log(`Found ${sentNoncesKeys.length} nonce-sent* keys`);
+
+  for (const sentNonceKey of sentNoncesKeys) {
+    const { chainId, walletAddress } = splitSentNoncesKey(sentNonceKey);
+
+    // Check blockchain for nonce
+    const rpcRequest = getRpcClient({
+      client: thirdwebClient,
+      chain: await getChain(chainId),
+    });
+
+    // The next unused nonce = transactionCount.
+    const transactionCount = await eth_getTransactionCount(rpcRequest, {
+      address: walletAddress,
+    });
+
+    // Get DB Nonce
+    const dbNonceCount = Number(
+      await redis.get(lastUsedNonceKey(chainId, walletAddress)),
+    );
+    job.log(
+      `[wallet ${walletAddress}] onchain Nonce: ${transactionCount} and DBNonce: ${dbNonceCount}`,
+    );
+    logger({
+      level: "debug",
+      message: `[nonceResyncWorker] onchain Nonce: ${transactionCount} and DBNonce: ${dbNonceCount}`,
+      service: "worker",
+    });
+
+    if (transactionCount >= dbNonceCount + 1) {
```

Comment:
Isn't this a bad condition that needs correcting?
The tx count is ahead of the redis nonce + 1. This means an increment will return an already used nonce.

Example:
- transactionCount = 5
- dbNonceCount = 2
- 5 is >= 2 + 1
- The next increments will be 3, 4, and 5 which are all already used nonces.

Do you plan to rely on the incr retries to fix this case? It seems like we could rebase the nonce here.

Reply:
this cron worker only handles the problem of chain nonce being stuck behind db nonce.
if the chain nonce is ahead of the db nonce, we should get a nonce too low, and we have a resolution in place within the sendTransaction worker to fix that